README.md
![Alt chrono](./.github/full-logo-with-tagline.png)

[![GitHub Release](https://img.shields.io/github/release-pre/gochrono/chrono.svg)](https://github.com/gochrono/chrono/releases)
[![Go Report Card](https://goreportcard.com/badge/github.com/gochrono/chrono)](https://goreportcard.com/report/github.com/gochrono/chrono)
[![Build Status](https://travis-ci.org/gochrono/chrono.svg?branch=master)](https://travis-ci.org/gochrono/chrono)
![GitHub](https://img.shields.io/github/license/gochrono/chrono.svg)
![Golang Version](https://img.shields.io/badge/Go-1.11-lightgrey.svg)



## Overview ##

Chrono is a time tracking tool written in Go.
It is fast and simple to use.

Want to know what you did with your time? Chrono will help you with that. Track how
long you spend on a project. Add notes so you know exactly what
you did.

[![asciicast](https://asciinema.org/a/pb5PlJHshi2gjpDeQYCj0gmfI.svg)](https://asciinema.org/a/pb5PlJHshi2gjpDeQYCj0gmfI)


## Installation ##

### Binaries Releases ###

To install the latest stable release:

``` bash
curl -L https://git.io/fpHAj | sh
```

We provide pre-built Chrono binaries for Windows, Linux, and macOS (Darwin) for x64 and i386 architectures.

If you'd like to install a specific version, checkout the [releases page](https://github.com/gochrono/chrono/releases).

If you want to run the absolute latest version, [build it from source](#building-chrono-from-the-source).

### Package Managers ###

If you are using __Arch Linux__, you can install the AUR package [chrono-git][2] to get the latest version or [chrono][3] for the latest stable release.

If you're using __Ubuntu__ or a __Debian__ based system, you can find debian packages (.deb) on the [releases page](https://github.com/gochrono/chrono/releases).

## Quickstart ##

### Usage ###

To start tracking time a project, use the `start` command:

``` bash
$ chrono start development +chrono
```

If you wanted to set the start time to something other than now, use the '--start' flag:

``` bash
$ chrono start development --at "2019-02-01 13:00"
```

Time is parsed by [jinzhu/now][now] so it should be in a format it can understand.

This creates a new __frame__ for the development project with the chrono tag.

Keep notes of what you do for a project with the `notes add` command:

``` bash
$ chrono notes add "made some awesome changes to the README"
$ chrono notes show
[0]: made some awesome changes to the README
```

The notes are added to the current __frame__.

Get information about the current frame with the `status` command:

``` bash
$ chrono status
Project development [chrono] started 10 seconds ago.
```
To stop tracking time for the current frame, use the `stop` command:

``` bash
$ chrono stop
Stopping project development [chrono], started 5 minutes ago (id: 073bbf).
```

You can show a chronolical list of the current day's session (or __frames__) through the `log` command:

``` bash
$ chrono log
Monday  3 December 2018
    (ID: 0d3131) 10:15 to 10:20     0h 05m 00s  development [chrono]
```

For a list of all available commands, use the `help` command:

```
$ chrono help
```

For a list of all available options and arguments for a command, use the `---help` flag:

```
$ chrono log --help
```

## Building Chrono from the Source ##

### Prequisite Tools ###

* [Git](https://git-scm.com/)
* [Go (at least Go 1.11)](https://golang.org/dl/)


### Downloading & building the source ###

Chrono uses [Go Modules](https://github.com/golang/go/wiki/Modules) to handle dependencies.

The easiest way to build from the source is to clone Chrono in a directory outsite of `GOPATH`, for example:

``` bash
mkdir ~/src && cd ~/src
git clone https://github.com/gochrono/chrono.git
cd chrono
go install
```

Make sure `$GOPATH/bin` is on `$PATH`.

If you haven't customized your `$GOPATH` then you'll need to add `~/go/bin` to `$PATH`.

## Contributing to Chrono ###

To contribute to the Chrono project or documentation, you should [fork the GitHub project](https://github.com/gochrono/chrono#fork-destination-box) and clone it to your machine.

For a complete guide to contributing to Chrono, see the [Contribution Guide](CONTRIBUTING.md).

We welcome contributions of many kinds from updating documentation, feature requests, bug reports & issues,
feature implementation, pull requests, answering other users questions, etc.

### Asking Support Questions ###

We currently don't have a discussion forum. For now, use the issue tracker to ask questions.

### Reporting Issues ###

If you believe you have found an issue or bad documentation, use
the GitHub issue tracker to report the problem to the Chrono maintainers.

When reporting an issue, please provide the version of chrono is use (`chrono version`)

[now]: https://github.com/jinzhu/now#parse-string-to-time
[1]: https://asciinema.org/a/pb5PlJHshi2gjpDeQYCj0gmfI
[2]: https://aur.archlinux.org/packages/chrono-git/
[3]: https://aur.archlinux.org/packages/chrono/


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2019-02-01 ##
### Added
- Added 'now' keyword to '--end' flag time string. Gets replaced with current time and date.
- Added '-d' flag to 'log' command that shows frames just for yesterday
- Added note in time format help docs that it should be wrapped in quotes.

### Fixed
- Fixed a bug where if the 'y' is capitalized when confirming a frame delete, it is treated as a no.
- Actually implemented frame update method, fixing issue where edit didn't do anything
- Added an actual error message rather than just crashing on an invalid tag in `start`
- Fixed the time format being shown in the command help as backwards


## [1.0.1] - 2018-12-26 ##
### Added
- Added `--show-notes` flag to `status` command. If passed to `status`, notes for the current frame will be shown
- Added check for existing project in `restart` command

### Fixed
- Fixed bug in `start` where if you started a new project with one already started, the first one would get overridden.

## [1.0.0] - 2018-12-19
### Added
- Added coloring to notes show command
- Updated help documentation for `start` and `notes` command
- Added `--at` flag to `stop` command to specify a time other than now for the new frame's end time.
- Added `--project` flag to `report` and `log` command to filter by project name
- Added `--tag` flag to `report` and `log` command to filter by tags
- Added `projects` command which shows a unique list of all project names used
- Added `restart` command which starts a new frame, using the last frame's project & tags
- Added `--version` flag which works the same as the `version` command
- Added `delete` command which deletes a saved frame by either an index or UUID
- Added `cancel` command which stops tracking the current frame without saving it
- Added `frames` command which shows a list of frame ID's
- Added global `--no-color` flag that doesn't print out ANSI color codes
- Added global `--verbose` flag that prints out helping debugging information
- Added config file support (config file should be located at ~/.config/chrono/config)


### Changed
- Renamed `--start` flag to `--at` in `start` command
- Frame UUID's are stored as strings instead of byte arrays now
- The state filename is now based on the format state.{storageType}
- The frames filename is now based on the format state.{storageType}

### Fixed
- Edge case where frames adjusted through the --round flag had negative times
- Bug where tags where stored with their ANSI color codes

## [1.0.0-beta] - 2018-10-11
### Added
- Core commands implemented
- Data storage format stabilized

[Unreleased]: https://github.com/gochrono/chrono/compare/v1.1.0...HEAD
[1.1.0]: https://github.com/gochrono/chrono/compare/v1.0.1...v1.1.0
[1.0.1]: https://github.com/gochrono/chrono/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/gochrono/chrono/compare/v1.0.0-beta...v1.0.0


commands/log.go
package commands

import (
	"fmt"
	"github.com/gochrono/chrono/chronolib"
	"github.com/jinzhu/now"
	"github.com/spf13/cobra"
	jww "github.com/spf13/jwalterweatherman"
	"os"
	"time"
)

var interval int
var logForCurrentWeek bool
var logForCurrentMonth bool
var logForCurrentYear bool
var logForAllTime bool
var logForYesterday bool
var logFrom string
var logTo string
var round bool
var logTags []string
var logProjects []string

// GetToFromTimespan is a helper function that takes in two time formats and attempts to parse them
func GetToFromTimespan(from string, to string) (chronolib.TimespanFilterOptions, error) {
	start := now.BeginningOfDay()
	end := now.EndOfDay()
	var err error
	if logTo != "" {
		start, err = ParseTime(logTo)
		if err != nil {
			fmt.Println("unable to parse time string")
			return chronolib.TimespanFilterOptions{}, err
		}
	}
	if logFrom != "" {
		end, err = ParseTime(logFrom)
		if err != nil {
			fmt.Println("unable to parse time string")
			return chronolib.TimespanFilterOptions{}, err
		}
	}
	return chronolib.TimespanFilterOptions{Start: start, End: end}, nil
}

func newLogCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "log",
		Short: "Get the frames logged for a single day",
		Long:  "Get the frames logged for a single day",
		Run: func(cmd *cobra.Command, args []string) {
			configDir := chronolib.GetCorrectConfigDirectory("")
			config := chronolib.GetConfig(configDir)

			if chronolib.ContainsMoreThanOneBooleanFlag(logForCurrentWeek, logForCurrentMonth, logForCurrentYear) {
				fmt.Println("Error: the folllowing flags are mutually exclusive: ['--week', '--year', '--month']")
				os.Exit(0)
			}

			var timespanFilterOptions chronolib.TimespanFilterOptions
			var err error
			if logTo != "" || logFrom != "" {
				timespanFilterOptions, err = GetToFromTimespan(logFrom, logTo)
				if err != nil {
					fmt.Println("unable to parse time string")
					return
				}
			} else {
				timespanFilterOptions = ParseTimespanFlags(TimespanFlags{
					AllTime:      logForAllTime,
					CurrentWeek:  logForCurrentWeek,
					CurrentMonth: logForCurrentMonth,
					CurrentYear:  logForCurrentYear,
					Yesterday:    logForYesterday,
				})
			}

			jww.INFO.Printf("timespan filter options: %v", timespanFilterOptions)

			frames, _ := chronolib.GetFrames(config)
			filteredFrames := frames.Filter(chronolib.FrameFilterOptions{
				TimespanFilter: timespanFilterOptions, Tags: logTags, Projects: logProjects,
			})
			timemap := chronolib.OrganizeFrameByTime(&filteredFrames)
			dates := chronolib.SortTimeMapKeys(&timemap)
			for _, date := range dates {
				fmt.Println(chronolib.FormatDateHeader(date))
				for _, frame := range timemap[date] {
					if round {
						newStartTime := GetAdjustedTime(frame.StartedAt)
						newEndTime := GetAdjustedTime(frame.EndedAt)
						if !chronolib.IsTimespanNegative(newStartTime, newEndTime) {
							frame.StartedAt = newStartTime
							frame.EndedAt = newEndTime
						}
						fmt.Println(chronolib.FormatFrameLine(frame))
					} else {
						fmt.Println(chronolib.FormatFrameLine(frame))
					}
					for _, note := range frame.Notes {
						fmt.Println(chronolib.FormatNoteLine(note))
					}
				}
			}
		},
	}
	cmd.Flags().BoolVarP(&logForCurrentWeek, "week", "w", false, "show frames for entire week")
	cmd.Flags().BoolVarP(&logForCurrentMonth, "month", "m", false, "show frames for entire month")
	cmd.Flags().BoolVarP(&logForCurrentYear, "year", "y", false, "show frames for entire year")
	cmd.Flags().BoolVarP(&logForYesterday, "yesterday", "d", false, "show frames for yesterday")
	cmd.Flags().BoolVarP(&logForAllTime, "all", "a", false, "show all frames")
	cmd.Flags().BoolVarP(&round, "round", "r", false, "round frames start and end times to the nearest interval (default: 5 mins)")
	cmd.Flags().StringVarP(&logFrom, "from", "f", "", "")
	cmd.Flags().StringVarP(&logTo, "to", "T", "", "")
	cmd.Flags().IntVarP(&interval, "interval", "i", 5, "the interval to round to in minutes")
	cmd.Flags().StringSliceVarP(&logTags, "tag", "t", []string{}, "only show frames that contain the given tag - can be used multiple times")
	cmd.Flags().StringSliceVarP(&logProjects, "project", "p", []string{}, "only show frames that contain the given project - can be used multiple times")
	cmd.Flags().BoolVarP(&verbose, "verbose", "v", false, "enable verbose output")
	return cmd
}

// GetAdjustedTime rounds a time to a given interval.
// E.g if the time is 15:02:13 and the interval is 5 minutes it
// would be rounded to 15:00:00
func GetAdjustedTime(t time.Time) time.Time {
	halfway := interval * 60 / 2
	rem := t.Minute()&interval*60 + t.Second()
	minutes := t.Minute() % interval
	if rem > halfway {
		if minutes == 0 {
			return t
		}
		return t.Add(time.Duration(interval-minutes) * time.Minute)
	}
	return t.Add(time.Duration(-minutes) * time.Minute)
}
